server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java

```diff
@@ -19,39 +19,178 @@
 
+import java.io.IOException;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.Iterator;
 import java.util.List;
+import java.util.Map;
 import java.util.Objects;
 import java.util.Optional;
 
+import javax.mail.Header;
 import javax.mail.MessagingException;
 
 import org.apache.james.core.MailAddress;
 import org.apache.mailet.Mail;
+import org.apache.mailet.PerRecipientHeaders;
 
 import com.github.steveash.guavate.Guavate;
+import com.google.common.collect.Multimap;
 
 public class MailDto {
-    public static MailDto fromMail(Mail mail) throws MessagingException {
+    public static MailDto fromMail(Mail mail, List<AdditionalFields> additionalFields) throws MessagingException, MissingRequestedField {
         return new MailDto(mail.getName(),
             Optional.ofNullable(mail.getSender()).map(MailAddress::asString),
             mail.getRecipients().stream().map(MailAddress::asString).collect(Guavate.toImmutableList()),
             Optional.ofNullable(mail.getErrorMessage()),
-            Optional.ofNullable(mail.getState()));
+            Optional.ofNullable(mail.getState()),
+            Optional.ofNullable(mail.getRemoteHost()),
+            Optional.ofNullable(mail.getRemoteAddr()),
+            Optional.ofNullable(mail.getLastUpdated()),
+            fetchAttributes(additionalFields, mail),
+            fetchPerRecipientsHeaders(additionalFields, mail),
+            fetchHeaders(additionalFields, mail),
+            fetchBody(additionalFields, mail),
+            fetchMessageSize(additionalFields, mail));
+    }
+
+    private static Optional<Long> fetchMessageSize(List<AdditionalFields> additionalFields, Mail mail) throws MissingRequestedField {
+        if (!additionalFields.contains(AdditionalFields.MESSAGE_SIZE)) {
+            return Optional.empty();
+        }
+        try {
+            return Optional.ofNullable(new Long(mail.getMessageSize()));
+        } catch (MessagingException e) {
+            throw new MissingRequestedField("messageSize");
+        }
+    }
+
+    private static Optional<String> fetchBody(List<AdditionalFields> additionalFields, Mail mail) throws MissingRequestedField {
+        if (!additionalFields.contains(AdditionalFields.BODY)) {
+            return Optional.empty();
+        }
+        try {
+            return Optional.ofNullable(mail.getMessage().getContent().toString());
+        } catch (IOException | MessagingException | NullPointerException e) {
+            throw new MissingRequestedField("body");
+        }
+    }
+
+    private static Optional<Map<String, List<String>>> fetchHeaders(List<AdditionalFields> additionalFields, Mail mail) throws MissingRequestedField {
+        if (!additionalFields.contains(AdditionalFields.HEADERS)) {
+            return Optional.empty();
+        }
+        Map<String, List<String>> headers = new HashMap<>();
+
+        Enumeration<Header> rawHeaders;
```

Comment:
we don't use variable declaration without initialization because it make reading harder by splitting your intent in two parts.